components/scream/src/physics/p3/p3_functions_find_impl.hpp

```diff
@@ -0,0 +1,97 @@
+#ifndef P3_FUNCTIONS_FIND_IMPL_HPP
+#define P3_FUNCTIONS_FIND_IMPL_HPP
+
+#include "p3_functions.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 find functions. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+Int Functions<S,D>
+::find_bottom (
+    const MemberType& team,
+    const ko::Unmanaged<view_1d<const Scalar> >& v, const Scalar& small,
+    const Int& kbot, const Int& ktop, const Int& kdir,
+    bool& log_present)
+{
+  log_present = false;
+  Int k_xbot = 0;
+  if (team.team_size() == 1) {
+    for (Int k = kbot; k != ktop + kdir; k += kdir) {
+      if (v(k) < small) continue;
+      k_xbot = k;
+      log_present = true;
+      break;
+    }
+  } else {
+    if (kdir == -1) {
+      Kokkos::parallel_reduce(
+        Kokkos::TeamThreadRange(team, kbot - ktop + 1), [&] (Int k_, int& lmax) {
```

Comment:
We are used to see either all 3 levels of parallelism (down to ThreadVectorRange) or some variation of Kokkos::single. Concerns are, obviously, for GPU builds. In this case, it's fine because there are no arithmetic operations, and the only operation happening is a min, which is a sort of 'fixed point'. It would be worth it to add a comment, so the user does not get puzzled about possible race conditions.

Reply:
We'll have to come up with a general pattern for communicating things like this. One possibility is to use `scream_kassert_msg` in every function that has expectations about Kokkos vector size.

Separately, and for clarification w.r.t. these comments, P3 always sets the Kokkos vector size to 1; all ||ism (for 128 levels, 128 threads = 4 warps).

Reply:
Ah ok. I was thinking about it, but I wasn't sure. That's another good solution; asserts are always verbose enough in explaining the culprit.